src/IO/H5/Helpers.cpp

```diff
@@ -339,24 +442,63 @@ bool contains_attribute(const hid_t file_id, const std::string& group_name,
          std::end(names);
 }
 
-DataVector read_data(const hid_t group_id, const std::string& dataset_name) {
-  // Read a DataVector from the group
-  const hid_t dataset_id =
-      H5Dopen2(group_id, dataset_name.c_str(), h5p_default());
-  CHECK_H5(dataset_id, "could not open dataset '" << dataset_name << "'");
-  // Get the number of points. These do not need a "close" call.
-  const hid_t space_id = H5Dget_space(dataset_id);
-  CHECK_H5(space_id, "Failed to open dataspace");
-  const hid_t number_of_points = H5Sget_simple_extent_npoints(space_id);
-  CHECK_H5(number_of_points, "Failed to get number of points");
-  H5Sclose(space_id);
-  // Load the data
-  DataVector data(static_cast<size_t>(number_of_points));
-  CHECK_H5(H5Dread(dataset_id, h5_type<double>(), h5s_all(), h5s_all(),
-                   h5p_default(), static_cast<void*>(data.data())),
-           "Failed to read data");
-  CHECK_H5(H5Dclose(dataset_id), "Failed to close dataset");
-  return data;
+hid_t open_dataset(const hid_t group_id,
+                   const std::string& dataset_name) noexcept {
+  const hid_t dataset_id = H5Dopen2(
+                           group_id, dataset_name.c_str(), h5p_default());
+  CHECK_H5(dataset_id, "Failed to open dataset '" << dataset_name << "'");
+  return dataset_id;
+}
+
+void close_dataset(const hid_t dataset_id) noexcept {
+CHECK_H5(H5Dclose(dataset_id), "Failed to close dataset");
+}
+
+hid_t open_dataspace(const hid_t dataset_id) noexcept {
+  const hid_t dataspace_id = H5Dget_space(dataset_id);
+  CHECK_H5(dataspace_id, "Failed to open dataspace");
+  return dataspace_id;
+}
+
+void close_dataspace(const hid_t dataspace_id) noexcept {
+  CHECK_H5(H5Sclose(dataspace_id), "Failed to close dataspace");
+}
+
+template <size_t Rank, typename T>
+T read_data(const hid_t group_id, const std::string& dataset_name) noexcept {
+  const hid_t dataset_id = open_dataset(group_id, dataset_name);
+  const hid_t dataspace_id = open_dataspace(dataset_id);
+  std::array<hsize_t, Rank == 0 ? 1 : Rank> size{}, max_size{};
```

Comment:
Allowing for arrays of size 0 causes problems in calls to the VectorToArray struct. For the rank 0 case the function `apply` is passed a vector of length `size`, and the first component of the vector is returned, so `size` cannot be 0

Reply:
`size` is an array, so I'm not sure exactly what you're trying to say.

The vector passed to `VectorToArray::apply` has length `size_product == 0 ? 1 : size_product`, which is always non-zero by construction, so that can't interfere with the indexing.  In fact, as far as I can see, the only reason the special check for zero `size_product` is needed is because you're making the `size` array longer than necessary when `Rank == 0`.